vsintegration/src/FSharp.Editor/CodeFixes/AddOpenCodeFixProvider.fs

```diff
@@ -2,40 +2,101 @@
 open CancellableTasks
 
 [<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.AddOpen); Shared>]
 type internal AddOpenCodeFixProvider [<ImportingConstructor>] (assemblyContentProvider: AssemblyContentProvider) =
     inherit CodeFixProvider()
 
+    static let br = Environment.NewLine
+
     let fixUnderscoresInMenuText (text: string) = text.Replace("_", "__")
 
     let qualifySymbolFix (context: CodeFixContext) (fullName, qualifier) =
-        context.RegisterFsharpFix(CodeFix.AddOpen, fixUnderscoresInMenuText fullName, [| TextChange(context.Span, qualifier) |])
-
-    let openNamespaceFix (context: CodeFixContext) ctx name ns multipleNames sourceText =
-        let displayText = "open " + ns + (if multipleNames then " (" + name + ")" else "")
-        let newText, _ = OpenDeclarationHelper.insertOpenDeclaration sourceText ctx ns
-        let changes = newText.GetTextChanges(sourceText)
+        {
+            Name = CodeFix.AddOpen
+            Message = fixUnderscoresInMenuText fullName
+            Changes = [ TextChange(context.Span, qualifier) ]
+        }
 
-        context.RegisterFsharpFix(CodeFix.AddOpen, fixUnderscoresInMenuText displayText, changes)
+    // Hey, I know what you're thinking: this is a horrible hack.
+    // Indeed it is, this is a better (but still bad) version of the OpenDeclarionHelper.
+    // The things should be actually fixed in the InsertionContext, it's bugged.
+    // But currently CompletionProvider also depends on InsertionContext and it's not tested enough.
+    // So fixing InsertionContext or OpenDeclarationHelper might break completion which would be bad.
+    // The hack below is at least heavily tested.
+    // And at least it shows what should be fixed down the line.
+    let getOpenDeclaration (sourceText: SourceText) (ctx: InsertionContext) (ns: string) =
+        // insertion context counts from 2, make the world sane
+        let insertionLineNumber = ctx.Pos.Line - 2
+        let margin = String(' ', ctx.Pos.Column)
+
+        let startLineNumber, openDeclaration =
+            match ctx.ScopeKind with
+            | ScopeKind.TopModule ->
+                match sourceText.Lines[ insertionLineNumber ].ToString().Trim() with
+
+                // explicit top level module
+                | line when line.StartsWith "module" && not (line.EndsWith "=") -> insertionLineNumber + 2, $"{margin}open {ns}{br}{br}"
+
+                // nested module, shouldn't be here
+                | line when line.StartsWith "module" -> insertionLineNumber, $"{margin}open {ns}{br}{br}"
+
+                // attribute, shouldn't be here
+                | line when line.StartsWith "[<" && line.EndsWith ">]" ->
+                    let moduleDeclLineNumber =
+                        sourceText.Lines
+                        |> Seq.skip insertionLineNumber
+                        |> Seq.findIndex (fun line -> line.ToString().Contains "module")
+                        |> (+) insertionLineNumber
```

Comment:
can this please be a normal lambda? Partially applied operators are hard to read, even if it's as simple as addition. We shouldn't really encourage it.

Reply:
Sure, there you go!